azurerm/resource_arm_batch_account.go

```diff
@@ -0,0 +1,224 @@
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	name := id.Path["batchAccounts"]
+	resourceGroupName := id.ResourceGroup
+
+	resp, err := client.Get(ctx, resourceGroupName, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			log.Printf("[DEBUG] Batch Account %q was not found in Resource Group %q - removing from state!", name, resourceGroupName)
+			return nil
+		}
+		return fmt.Errorf("Error reading the state of Batch account %q: %+v", name, err)
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("resource_group_name", resourceGroupName)
+	d.Set("pool_allocation_mode", resp.PoolAllocationMode)
+
+	if location := resp.Location; location != nil {
+		d.Set("location", azureRMNormalizeLocation(*location))
+	}
+
+	if resp.AutoStorage != nil {
+		d.Set("storage_acount_id", resp.AutoStorage.StorageAccountID)
+	}
+
+	flattenAndSetTags(d, resp.Tags)
+
+	return nil
+}
+
+func resourceArmBatchAccountUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).batchAccountClient
+	ctx := meta.(*ArmClient).StopContext
+
+	log.Printf("[INFO] preparing arguments for Azure Batch account update.")
+
+	resourceGroupName := d.Get("resource_group_name").(string)
+	name := d.Get("name").(string)
+
+	storageAccountId := d.Get("storage_account_id").(string)
+	tags := d.Get("tags").(map[string]interface{})
+
+	parameters := batch.AccountUpdateParameters{
+		AccountUpdateProperties: &batch.AccountUpdateProperties{
+			AutoStorage: &batch.AutoStorageBaseProperties{
+				StorageAccountID: &storageAccountId,
+			},
+		},
+		Tags: expandTags(tags),
+	}
+
+	_, err := client.Update(ctx, resourceGroupName, name, parameters)
+	if err != nil {
+		return fmt.Errorf("Error updating Batch account %q (Resource Group %q): %+v", name, resourceGroupName, err)
+	}
+
+	read, err := client.Get(ctx, resourceGroupName, name)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Batch account %q (Resource Group %q): %+v", name, resourceGroupName, err)
+	}
+
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read Batch account %q (resource group %q) ID", name, resourceGroupName)
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmBatchAccountRead(d, meta)
+}
+
+func resourceArmBatchAccountDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).batchAccountClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	name := id.Path["batchAccounts"]
+	resourceGroupName := id.ResourceGroup
+
+	future, err := client.Delete(ctx, resourceGroupName, name)
+	if err != nil {
+		return fmt.Errorf("Error deleting Batch account %q (Resource Group %q): %+v", name, resourceGroupName, err)
+	}
+
+	err = future.WaitForCompletionRef(ctx, client.Client)
+	// if the error is not that the Batch account was not found
+	if err != nil && future.Response().StatusCode != http.StatusNotFound {
```

Comment:
if the connections reset `future.Response()` is nil so there's a potential panic here - could we switch to using the helper method which handles this instead:

```
if err != nil {
  if !response.WasNotFound(future.Response()) {
    return fmt.Errorf("Error waiting for deletion of Batch account %q (Resource Group %q): %+v", name, resourceGroupName, err)
  }
}
```

Reply:
Same

```
if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
  if !response.WasNotFound(future.Response()) {
    return fmt.Errorf("Error waiting for deletion of Batch account %q (Resource Group %q): %+v", name, resourceGroupName, err)
  }
}
```

Reply:
Done.